### ISSUE TITLE
New testing operator for data examples

### ISSUE THREAD

**Issue by blerner** [brownplt/pyret-lang#1218]
Brainstorm idea: what if we added a new testing operator,

```
<expr> is-a <ann>
```

with semantics: "Try to construct the expression.  If any annotation-failure occurs, raise the error 'Attempting to construct this value was internally inconsistent, because <subexpr> failed <sub-annotation>'.  If it succeeds, check if its result matches the annotation.  If that fails, the test fails with 'The given expression did not have the expected type.'"

Obvious limitation: This fails for polymorphic data or for functions-as-data, because we don't do deep nesting.  But it might get us a large part of the way to a "check-expect for data"...

Thoughts?  What other limitations or benefits do you see?  Is it worth it?

**Comment by blerner** [comment]
(ping @shriram and @jpolitz for input)

**Comment by sorawee** [comment]
Should `eval("(lam x x)") is-a v-fun` work too?

**Comment by blerner** [comment]
Sure, why not?  The usefulness of data examples is highest when you're a complete novice (and `eval` doesn't exist yet) and when you're an expert (and not a showoff).  In the middle, hukairs?

**Comment by jpolitz** [comment]
Does that make the position after "is-a" be an expression position?

Or is it that variant constructors are also acceptable as annotations, and
the is-a form figures it out?

**Comment by blerner** [comment]
Oh sorry, I totally missed that in my initial reading of the comment; I was focused on the `eval` part.  No, I'm imagining it should be an `Ann`, so `... is-a Value%(is-v-fun)` would be ok, but `is-a v-fun` isn't.  But I think I'm ok with that -- typically in Fundies we say "Give us examples of a Foobar", so writing `eval("(lam x x)") is-a Value` would be entirely appropriate.  How key is it, in CS19 or elsewhere, to allow writing refinement annotations as examples of data?  (The closest we get in Fundies is "non-empty lists of X", so I'd need `List%(is-link)`, maybe.)

**Comment by sorawee** [comment]
Couldn't we do this instead?

```
<expr> is-a <literal-id>
```

and before sending this to well-formedness checker, transform the code to

```
<expr> satisfies is-<literal-id>
```

One draw back is the error message in case `is-<literal-id>` is unbound.

**Comment by jpolitz** [comment]
I like the idea of an annotation, because I like the idea that the type
checker could be involved in testing. What if it was a unit test error if
you misunderstood types, instead of refusing to run the program? Could be
cool for testing functions that return functions by describing their type.

**Comment by blerner** [comment]
Yeah, I'd prefer standard annotations over unhygienic name manipulation as the core mechanic of this testing form. That helps us avoid having to explain another special-purpose mechanism...

**Comment by shriram** [comment]
I don't like the name manipulation. These kinds of tricks invariably come
back to hurt.

1. I'm frankly not quite sure what this feature is for.

2. Wouldn't `satisfies` achieve essentially the same thing? What am I
missing?

**Comment by blerner** [comment]
The DR for functions says, "Come up with a signature & purpose, then come up with functional examples of what you think the function should do".  We write those functional examples as `foo(bar) is baz`.  The DR for data definitions says, "Come up with a definition and interpretation, then come up with examples of your data".  We currently have no way of writing such examples down _as test cases_.  

* Sure, we could just write down the data as top-level statements, but then they print out in the REPL as part of the execution of the program.  And, they tell you nothing about whether those values are really part of the data definition you expected.

* Sure, we could write `link(5, empty) satisfies is-List`, but that's (IMO) ugly and unintuitive -- saying that's a fully satisfactory approach is analogous to saying we don't need `is`, since we could write `foo(bar) satisfies equal-always(_, baz)`.  Writing `link(3, empty) is-a List` reads better and is pleasingly parallel to writing functional examples.

This isn't semantically necessary, but I think it's ergonomically better support for a part of the DR that we currently skimp on.

**Comment by blerner** [comment]
(I'm imagining this might be very useful for giving `data` definitions their own `where` blocks.)

**Comment by shriram** [comment]
Got it. I like it.

But if we're taking this part of the DR seriously, there's a *much more
serious* issue to consider. We need to be able to define instances of data
in the `where` block of a data definition, then use those instances in the
`where` blocks of functions, instead of being forced to push all those to
the top-level. (There's probably a request for this from 3 years ago…)

**Comment by blerner** [comment]
Yes, that's an issue.  It means you're asking us to give up lexical scoping for Pyret, though.  It means thinking *extremely* carefully about whether those where-block definitions get exported when we export data definitions, whether they even can be exported.

I *don't* see my particular proposal as suggesting we allow new binding statements in some suggested where-block for data, though.  That feels like a wholly new thing.

**Comment by shriram** [comment]
I think it's a different notion of lexical scoping. To me, it's not
inconsistent with lexical scoping.

But yes, your proposal is different. I'm just pointing out that since we
have a motivation to make this "work well for the DR", why not address
another long-standing problem that also falls under that rubric.

**Comment by sorawee** [comment]
We kinda give up lexical scope already, no?

```
a = 1
check:
  a is 0
end

shadow a = 2
```

will result in a testing error that 1 (not 2!) is not equal 0.

**Comment by blerner** [comment]
I don't see the problem with lexical scoping there: the test uses the scope in which it was written, and the shadow declaration doesn't exist yet.

Shriram's suggestion about allowing definitions in one check block to be visible in a different check block, however, is definitely at odds with the traditional "scope nesting follows lexical nesting" we've tried to maintain.  If we add a new notion that check blocks introduce a new kind of scope, that can be added to later (!), that seems like either the kind of scope hacking that's done in Racket to support macros that introduce bindings (which always struck me as a bit gross), or the kinds of weird scopes that we disparaged in the lambda-python paper...

I don't have a good mental model of how Shriram's requested scopes should work (in terms of, do all check blocks see each other? How should name resolution work? What order do the blocks execute in, if they can see each other's definitions? Can those definitions be visible across module boundaries? What about generativity of check blocks nested within functions?) so I'm not sure what to think about that yet.
